cohttp-lwt/src/connection.ml

```diff
@@ -97,8 +97,8 @@ struct
 
   let rec reader connection =
     match connection.state with
-    | Connecting _ -> assert false
-    | Closed | Failed _ -> assert false
+    | Connecting _ | Failed _ -> assert false
```

Comment:
If it may trigger, we should actually raise an exception instead

Reply:
I don't believe it can trigger. `reader` is the only function setting `state` to `Failed _`. `writer` will only close the out channel, so that the `reader` may still receive responses. If this assertion still does trigger I'd like to know that and fix my understanding of the code.

Reply:
One more reason to raise a meaningful exception, imho, to make sure that we can track down where it comes from. Assert false are usually a nightmare to pinpoint

Reply:
This is how it failed on me:
```
Fatal error: exception File "cohttp-lwt/src/connection.ml", line 104, characters 16-22: Assertion failed
Raised at Cohttp_lwt__Connection.Make.reader in file "cohttp-lwt/src/connection.ml", line 104, characters 16-28
Called from Lwt.Sequential_composition.bind.create_result_promise_and_callback_if_deferred.callback in file "src/core/lwt.ml", line 1860,
characters 23-26
```
In that case it was pretty specific, which assertion failed. So I don't really get what you are missing here. What do you propose?

Reply:
Oh, ok this is much more informative than I remember. Was this improved recently? In this case, let's leave it like that!